spt/features/vag_trace.cpp

```diff
@@ -0,0 +1,273 @@
+{
+	Vector transition = target - enter_origin;
+	Vector enter[3];
+	AngleVectors(enter_angles, &enter[0], &enter[1], &enter[2]);
+	enter[0] *= -1;
+	enter[1] *= -1;
+
+	float det = enter[0][0] * (enter[1][1] * enter[2][2] - enter[2][1] * enter[1][2])
+	            - enter[0][1] * (enter[1][0] * enter[2][2] - enter[1][2] * enter[2][0])
+	            + enter[0][2] * (enter[1][0] * enter[2][1] - enter[1][1] * enter[2][0]);
+	float invdet;
+	if (det == 0)
+	{
+		invdet = 1;
+	}
+	else
+	{
+		invdet = 1 / det;
+	}
+
+	Vector inv_enter[3];
+	inv_enter[0][0] = (enter[1][1] * enter[2][2] - enter[2][1] * enter[1][2]) * invdet;
+	inv_enter[0][1] = (enter[0][2] * enter[2][1] - enter[0][1] * enter[2][2]) * invdet;
+	inv_enter[0][2] = (enter[0][1] * enter[1][2] - enter[0][2] * enter[1][1]) * invdet;
+	inv_enter[1][0] = (enter[1][2] * enter[2][0] - enter[1][0] * enter[2][2]) * invdet;
+	inv_enter[1][1] = (enter[0][0] * enter[2][2] - enter[0][2] * enter[2][0]) * invdet;
+	inv_enter[1][2] = (enter[1][0] * enter[0][2] - enter[0][0] * enter[1][2]) * invdet;
+	inv_enter[2][0] = (enter[1][0] * enter[2][1] - enter[2][0] * enter[1][1]) * invdet;
+	inv_enter[2][1] = (enter[2][0] * enter[0][1] - enter[0][0] * enter[2][1]) * invdet;
+	inv_enter[2][2] = (enter[0][0] * enter[1][1] - enter[1][0] * enter[0][1]) * invdet;
+
+	Vector exit_portal_coords;
+	exit_portal_coords.x =
+	    transition.x * inv_enter[0][0] + transition.y * inv_enter[1][0] + transition.z * inv_enter[2][0];
+	exit_portal_coords.y =
+	    transition.x * inv_enter[0][1] + transition.y * inv_enter[1][1] + transition.z * inv_enter[2][1];
+	exit_portal_coords.z =
+	    transition.x * inv_enter[0][2] + transition.y * inv_enter[1][2] + transition.z * inv_enter[2][2];
+
+	Vector exit[3];
+	AngleVectors(exit_angles, &exit[0], &exit[1], &exit[2]);
+
+	det = exit[0][0] * (exit[1][1] * exit[2][2] - exit[2][1] * exit[1][2])
+	      - exit[0][1] * (exit[1][0] * exit[2][2] - exit[1][2] * exit[2][0])
+	      + exit[0][2] * (exit[1][0] * exit[2][1] - exit[1][1] * exit[2][0]);
+	if (det == 0)
+	{
+		invdet = 1;
+	}
+	else
+	{
+		invdet = 1 / det;
+	}
+
+	Vector inv_exit_angles[3];
+	inv_exit_angles[0][0] = (exit[1][1] * exit[2][2] - exit[2][1] * exit[1][2]) * invdet;
+	inv_exit_angles[0][1] = (exit[0][2] * exit[2][1] - exit[0][1] * exit[2][2]) * invdet;
+	inv_exit_angles[0][2] = (exit[0][1] * exit[1][2] - exit[0][2] * exit[1][1]) * invdet;
+	inv_exit_angles[1][0] = (exit[1][2] * exit[2][0] - exit[1][0] * exit[2][2]) * invdet;
+	inv_exit_angles[1][1] = (exit[0][0] * exit[2][2] - exit[0][2] * exit[2][0]) * invdet;
+	inv_exit_angles[1][2] = (exit[1][0] * exit[0][2] - exit[0][0] * exit[1][2]) * invdet;
+	inv_exit_angles[2][0] = (exit[1][0] * exit[2][1] - exit[2][0] * exit[1][1]) * invdet;
+	inv_exit_angles[2][1] = (exit[2][0] * exit[0][1] - exit[0][0] * exit[2][1]) * invdet;
+	inv_exit_angles[2][2] = (exit[0][0] * exit[1][1] - exit[1][0] * exit[0][1]) * invdet;
```

Comment:
if these matrix operations do something simple like calculate the inverse, there's probably an SDK function for it